# Benjamin Freund
# Week 10 Assignment

## Task 1

The first task in this assignment required me to recreate a given dataset as a CSV and upload it to my GitHub repository. This copy had to have all of the same spacing as the original. The CSV was uploaded to my GitHub repository, and a link to the raw data can be found in the first code cell of the second task.

## Task 2

The second task asked me to transform the data into a tidy dataset. This required many steps. First, I imported the Pandas library, read the CSV into the data variable, and displayed the data variable.

In [1]:
# Importing the Pandas library
import pandas as pd

# Reading the CSV into the data variable
data = pd.read_csv('https://raw.githubusercontent.com/freundb3/AIM-5001/master/B_Freund_Week_10_Data.csv')

# Displaying data
data

Unnamed: 0 Unnamed: 1  Los Angeles Phoenix  San Diego  San Francisco Seattle
0     ALASKA    on time        497.0     221      212.0          503.0   1,841
1        NaN    delayed         62.0      12       20.0          102.0     305
2        NaN        NaN          NaN     NaN        NaN            NaN     NaN
3    AM WEST    on time        694.0   4,840      383.0          320.0     201
4        NaN    delayed        117.0     415       65.0          129.0      61

As one can see, the initial dataset is very messy. There are lots of NaNs, unnamed columns, and variables as column names. Overall, this is not a tidy dataset.

The first step to tidy this data is to rename the unnamed columns. I will call the first column Airline and the second Status.

In [2]:
# Reassigning column names and storing the new dataset in new_columns
new_columns = data.rename(columns = {'Unnamed: 0':'Airline', 'Unnamed: 1':'Status'})

# Displaying new_columns
new_columns

Airline   Status  Los Angeles Phoenix  San Diego  San Francisco Seattle
0   ALASKA  on time        497.0     221      212.0          503.0   1,841
1      NaN  delayed         62.0      12       20.0          102.0     305
2      NaN      NaN          NaN     NaN        NaN            NaN     NaN
3  AM WEST  on time        694.0   4,840      383.0          320.0     201
4      NaN  delayed        117.0     415       65.0          129.0      61

Next, I will add in the missing values in the Airline column. I know from looking at the original dataset that the airline at index 1 should be ALASKA and the airline at index 2 should be AM WEST. I will add those values in accordingly.

In [3]:
# Inserting the airline values at their appropriate positions
new_columns.at[1, 'Airline'] = 'ALASKA'
new_columns.at[4, 'Airline'] = 'AM WEST'

# Displaying new_columns
new_columns

Airline   Status  Los Angeles Phoenix  San Diego  San Francisco Seattle
0   ALASKA  on time        497.0     221      212.0          503.0   1,841
1   ALASKA  delayed         62.0      12       20.0          102.0     305
2      NaN      NaN          NaN     NaN        NaN            NaN     NaN
3  AM WEST  on time        694.0   4,840      383.0          320.0     201
4  AM WEST  delayed        117.0     415       65.0          129.0      61

Now we have a row that is completely full of NaNs. This row was a blank row in the original dataset, so I can get rid of it now.

In [4]:
# Dropping the row with index 2 and storing the new dataset in new_rows
new_rows = new_columns.drop(2)

# Displaying new_rows
new_rows

Airline   Status  Los Angeles Phoenix  San Diego  San Francisco Seattle
0   ALASKA  on time        497.0     221      212.0          503.0   1,841
1   ALASKA  delayed         62.0      12       20.0          102.0     305
3  AM WEST  on time        694.0   4,840      383.0          320.0     201
4  AM WEST  delayed        117.0     415       65.0          129.0      61

This looks great initially, but, upon closer inspection, one can see that the index numbers did not reset (i.e. the 2 is simply missing). I now need to reset the index numbers.

In [5]:
# Resetting the index numbers and storing it in reset_index
reset_index = new_rows.reset_index(drop = True)

# Displaying reset_index
reset_index

Airline   Status  Los Angeles Phoenix  San Diego  San Francisco Seattle
0   ALASKA  on time        497.0     221      212.0          503.0   1,841
1   ALASKA  delayed         62.0      12       20.0          102.0     305
2  AM WEST  on time        694.0   4,840      383.0          320.0     201
3  AM WEST  delayed        117.0     415       65.0          129.0      61

Now, I need to pivot the dataset from wide format to long format. To do this, I need to use the melt function and specify the grouping columns, Airline and Status.

In [6]:
# Pivoting the dataset and saving it in reshaped
reshaped = pd.melt(reset_index, id_vars = ['Airline', 'Status'])

# Displaying reshaped
reshaped

Airline   Status       variable  value
0    ALASKA  on time    Los Angeles    497
1    ALASKA  delayed    Los Angeles     62
2   AM WEST  on time    Los Angeles    694
3   AM WEST  delayed    Los Angeles    117
4    ALASKA  on time        Phoenix    221
5    ALASKA  delayed        Phoenix     12
6   AM WEST  on time        Phoenix  4,840
7   AM WEST  delayed        Phoenix    415
8    ALASKA  on time      San Diego    212
9    ALASKA  delayed      San Diego     20
10  AM WEST  on time      San Diego    383
11  AM WEST  delayed      San Diego     65
12   ALASKA  on time  San Francisco    503
13   ALASKA  delayed  San Francisco    102
14  AM WEST  on time  San Francisco    320
15  AM WEST  delayed  San Francisco    129
16   ALASKA  on time        Seattle  1,841
17   ALASKA  delayed        Seattle    305
18  AM WEST  on time        Seattle    201
19  AM WEST  delayed        Seattle     61

Now the dataset is in long format, but the column names are still not correct. I will now change the variable column to Destination and the value column to Number of Flights.

In [7]:
# Reassigning column names and storing the new dataset in reshaped_names
reshaped_named = reshaped.rename(columns = {'variable':'Destination', 'value':'Number of Flights'})

# Displaying reshaped_names
reshaped_named

Airline   Status    Destination Number of Flights
0    ALASKA  on time    Los Angeles               497
1    ALASKA  delayed    Los Angeles                62
2   AM WEST  on time    Los Angeles               694
3   AM WEST  delayed    Los Angeles               117
4    ALASKA  on time        Phoenix               221
5    ALASKA  delayed        Phoenix                12
6   AM WEST  on time        Phoenix             4,840
7   AM WEST  delayed        Phoenix               415
8    ALASKA  on time      San Diego               212
9    ALASKA  delayed      San Diego                20
10  AM WEST  on time      San Diego               383
11  AM WEST  delayed      San Diego                65
12   ALASKA  on time  San Francisco               503
13   ALASKA  delayed  San Francisco               102
14  AM WEST  on time  San Francisco               320
15  AM WEST  delayed  San Francisco               129
16   ALASKA  on time        Seattle             1,841
17   ALASKA  delayed        Seattle               305
18  AM WEST  on time        Seattle               201
19  AM WEST  delayed        Seattle                61

Everything in the dataset is now correct. However, I want to get the dataset to be in a similar order as the original dataset. Therefore, I will order the dataset first by Airline, then by Status, and then by Destination.

In [8]:
# Ordering the dataset by Airline, Status, and Destination and storing it in ordered
ordered = reshaped_named.sort_values(['Airline', 'Status', 'Destination'])

# Resetting the index numbers and storing it in ordered_indexed
ordered_indexed = ordered.reset_index(drop = True)

# Removing commas from entries with commas
ordered_indexed['Number of Flights'][9] = 1841
ordered_indexed['Number of Flights'][16] = 4840

# Changing all of the values in the Number of Flights column to integers
ordered_indexed['Number of Flights'] = ordered_indexed['Number of Flights'].astype(int)

# Displaying ordered_indexed
ordered_indexed

Airline   Status    Destination  Number of Flights
0    ALASKA  delayed    Los Angeles                 62
1    ALASKA  delayed        Phoenix                 12
2    ALASKA  delayed      San Diego                 20
3    ALASKA  delayed  San Francisco                102
4    ALASKA  delayed        Seattle                305
5    ALASKA  on time    Los Angeles                497
6    ALASKA  on time        Phoenix                221
7    ALASKA  on time      San Diego                212
8    ALASKA  on time  San Francisco                503
9    ALASKA  on time        Seattle               1841
10  AM WEST  delayed    Los Angeles                117
11  AM WEST  delayed        Phoenix                415
12  AM WEST  delayed      San Diego                 65
13  AM WEST  delayed  San Francisco                129
14  AM WEST  delayed        Seattle                 61
15  AM WEST  on time    Los Angeles                694
16  AM WEST  on time        Phoenix               4840
17  AM WEST  on time      San Diego                383
18  AM WEST  on time  San Francisco                320
19  AM WEST  on time        Seattle                201

The dataset has now been transformed into a tidy dataset.

## Task 3

The next task is to answer the following two questions:

1. For each city, which airline had the best on time performance?
2. Which airline had the best overall on time performance?

I will look at each question individually.

### For each city, which airline had the best on time performance?

To answer this question, I need to calculate the total number of flights with each airline from each city, and then calculate (on time) / (total). The calculations are below.

#### Los Angeles

In [9]:
# Getting all flights where the destination is LA and storing them in LA
LA = ordered_indexed[ordered_indexed['Destination'] == 'Los Angeles']

# Getting all flights where the destination is LA and the airline is Alaska and storing them in LA_Alaska
LA_Alaska = LA[LA['Airline'] == 'ALASKA']

# Getting the percentage of on time flights to LA on the Alaska airline
(LA_Alaska[LA_Alaska['Status'] == 'on time']['Number of Flights'])/(sum(LA_Alaska['Number of Flights']))

5    0.889088
Name: Number of Flights, dtype: float64

In [10]:
# Getting all flights where the destination is LA and the airline is AM West and storing them in LA_Amwest
LA_Amwest = LA[LA['Airline'] == 'AM WEST']

# Getting the percentage of on time flights to LA on the AM West airline
(LA_Amwest[LA_Amwest['Status'] == 'on time']['Number of Flights'])/(sum(LA_Amwest['Number of Flights']))

15    0.855734
Name: Number of Flights, dtype: float64

For Los Angeles, 88.91% of Alaska's flights were on time, while 85.57% of AM West's flights were on time. Therefore, Alaska had the best on time performance.

#### Phoenix

In [11]:
# Getting all flights where the destination is Phoenix and storing them in phoenix
phoenix = ordered_indexed[ordered_indexed['Destination'] == 'Phoenix']

# Getting all flights where the destination is Phoenix and the airline is Alaska and storing them in phoenix_Alaska
phoenix_Alaska = phoenix[phoenix['Airline'] == 'ALASKA']

# Getting the percentage of on time flights to Phoenix on the Alaska airline
(int(phoenix_Alaska[phoenix_Alaska['Status'] == 'on time']['Number of Flights']))/(sum((int(phoenix_Alaska['Number of Flights'][1]), int(phoenix_Alaska['Number of Flights'][6]))))

0.9484978540772532

In [12]:
# Getting all flights where the destination is Phoenix and the airline is AM West and storing them in phoenix_Amwest
phoenix_Amwest = phoenix[phoenix['Airline'] == 'AM WEST']

# Getting the percentage of on time flights to Phoenix on the AM West airline
(int(phoenix_Amwest[phoenix_Amwest['Status'] == 'on time']['Number of Flights']))/(sum((int(phoenix_Amwest['Number of Flights'][11]), int(phoenix_Amwest['Number of Flights'][16]))))

0.9210275927687916

For Phoenix, 94.85% of Alaska's flights were on time, while 92.10% of AM West's flights were on time. Therefore, Alaska had the best on time performance.

#### San Diego

In [13]:
# Getting all flights where the destination is SD and storing them in SD
SD = ordered_indexed[ordered_indexed['Destination'] == 'San Diego']

# Getting all flights where the destination is SD and the airline is Alaska and storing them in SD_Alaska
SD_Alaska = SD[SD['Airline'] == 'ALASKA']

# Getting the percentage of on time flights to SD on the Alaska airline
(SD_Alaska[SD_Alaska['Status'] == 'on time']['Number of Flights'])/(sum(SD_Alaska['Number of Flights']))

7    0.913793
Name: Number of Flights, dtype: float64

In [14]:
# Getting all flights where the destination is SD and the airline is AM West and storing them in SD_Amwest
SD_Amwest = SD[SD['Airline'] == 'AM WEST']

# Getting the percentage of on time flights to SD on the AM West airline
(SD_Amwest[SD_Amwest['Status'] == 'on time']['Number of Flights'])/(sum(SD_Amwest['Number of Flights']))

17    0.854911
Name: Number of Flights, dtype: float64

For San Diego, 91.38% of Alaska's flights were on time, while 85.49% of AM West's flights were on time. Therefore, Alaska had the best on time performance.

#### San Francisco

In [15]:
# Getting all flights where the destination is SF and storing them in SF
SF = ordered_indexed[ordered_indexed['Destination'] == 'San Francisco']

# Getting all flights where the destination is SF and the airline is Alaska and storing them in SF_Alaska
SF_Alaska = SF[SF['Airline'] == 'ALASKA']

# Getting the percentage of on time flights to SF on the Alaska airline
(SF_Alaska[SF_Alaska['Status'] == 'on time']['Number of Flights'])/(sum(SF_Alaska['Number of Flights']))

8    0.831405
Name: Number of Flights, dtype: float64

In [16]:
# Getting all flights where the destination is SF and the airline is AM West and storing them in SF_Amwest
SF_Amwest = SF[SF['Airline'] == 'AM WEST']

# Getting the percentage of on time flights to SF on the AM West airline
(SF_Amwest[SF_Amwest['Status'] == 'on time']['Number of Flights'])/(sum(SF_Amwest['Number of Flights']))

18    0.712695
Name: Number of Flights, dtype: float64

For San Francisco, 83.14% of Alaska's flights were on time, while 71.27% of AM West's flights were on time. Therefore, Alaska had the best on time performance.

#### Seattle

In [17]:
# Getting all flights where the destination is Seattle and storing them in seattle
seattle = ordered_indexed[ordered_indexed['Destination'] == 'Seattle']

# Getting all flights where the destination is Seattle and the airline is Alaska and storing them in seattle_Alaska
seattle_Alaska = seattle[seattle['Airline'] == 'ALASKA']

# Getting the percentage of on time flights to Seattle on the Alaska airline
(int(seattle_Alaska[seattle_Alaska['Status'] == 'on time']['Number of Flights']))/(sum((int(seattle_Alaska['Number of Flights'][4]), int(seattle_Alaska['Number of Flights'][9]))))

0.8578751164958062

In [18]:
# Getting all flights where the destination is Seattle and the airline is AM West and storing them in seattle_Amwest
seattle_Amwest = seattle[seattle['Airline'] == 'AM WEST']

# Getting the percentage of on time flights to Seattle on the AM West airline
(int(seattle_Amwest[seattle_Amwest['Status'] == 'on time']['Number of Flights']))/(sum((int(seattle_Amwest['Number of Flights'][14]), int(seattle_Amwest['Number of Flights'][19]))))

0.767175572519084

For Seattle, 85.79% of Alaska's flights were on time, while 76.72% of AM West's flights were on time. Therefore, Alaska had the best on time performance.

For every city, Alaska had the best on time performance.

### Which airline had the best overall on time performance?

To answer this question, I need to calculate (on time) / (total) for each airline. The calculations are below.

In [19]:
# Getting all flights where the airline is Alaska and storing them in Alaska
Alaska = ordered_indexed[ordered_indexed['Airline'] == 'ALASKA']

# Getting all on time flights where the airline is Alaska and storing them in ontime_alaska
ontime_alaska = Alaska[Alaska['Status'] == 'on time']

# Getting the percentage of on time flights on the Alaska airline
(ontime_alaska['Number of Flights'].sum())/(Alaska['Number of Flights'].sum())

0.8672847682119206

In [20]:
# Getting all flights where the airline is AM West and storing them in Amwest
Amwest = ordered_indexed[ordered_indexed['Airline'] == 'AM WEST']

# Getting all on time flights where the airline is AM West and storing them in ontime_amwest
ontime_amwest = Amwest[Amwest['Status'] == 'on time']

# Getting the percentage of on time flights on the AM West airline
(ontime_amwest['Number of Flights'].sum())/(Amwest['Number of Flights'].sum())

0.8910726643598615

86.73% of Alaska's flights were on time, while 89.11% of AM West's flights were on time. Therefore, AM West had the best on time performance.

## Task 4

The fourth task asks what changes I would make if I were to transform the dataset back from a long format to a wide format. It is important to note that I would not keep the original format. The original format makes it difficult to transform the dataset between wide and long formats, as evidenced in task 2.

Rather, to return the dataset to a wide format, I would simply use the unstack function twice. The unstack function transforms a dataset from long format to wide format, while the stack function does the opposite. This will allow for easier future transitions between the two formats.

In [21]:
# Changing the dataframe from a long format to a wide format
pd.DataFrame(ordered_indexed.unstack().unstack())

0        1          2              3        4  \
Airline                 ALASKA   ALASKA     ALASKA         ALASKA   ALASKA   
Status                 delayed  delayed    delayed        delayed  delayed   
Destination        Los Angeles  Phoenix  San Diego  San Francisco  Seattle   
Number of Flights           62       12         20            102      305   

                             5        6          7              8        9  \
Airline                 ALASKA   ALASKA     ALASKA         ALASKA   ALASKA   
Status                 on time  on time    on time        on time  on time   
Destination        Los Angeles  Phoenix  San Diego  San Francisco  Seattle   
Number of Flights          497      221        212            503     1841   

                            10       11         12             13       14  \
Airline                AM WEST  AM WEST    AM WEST        AM WEST  AM WEST   
Status                 delayed  delayed    delayed        delayed  delayed   
Destination        Los Angeles  Phoenix  San Diego  San Francisco  Seattle   
Number of Flights          117      415         65            129       61   

                            15       16         17             18       19  
Airline                AM WEST  AM WEST    AM WEST        AM WEST  AM WEST  
Status                 on time  on time    on time        on time  on time  
Destination        Los Angeles  Phoenix  San Diego  San Francisco  Seattle  
Number of Flights          694     4840        383            320      201